Improving shots estimate to give errors within user-specified precision.

jsaroni opened the pull request:
### Summary 
New PR to replace #12618 because that one was from an incorrect branch.

Fixes #issue-12414 (https://github.com/Qiskit/qiskit/issues/12414). When the user specifies the precision in EstimatorV2.run, the actual standard deviation from results is not bounded by or close to the user-specified precision for an observable that is a general linear combination of Pauli strings. This pull request gives a more accurate approximation to the number of shots needed for the desired precision for such observables. The tests to verify this are based on https://docs.quantum.ibm.com/api/qiskit-ibm-runtime/qiskit_ibm_runtime.EstimatorV2 and https://github.com/Qiskit/qiskit-ibm-runtime/issues/1684.

### Details and comments
The shots estimation accounts for the sum of the absolute value squared of the coefficients of the observable as proposed in https://github.com/Qiskit/qiskit/issues/12414 to give a standard deviation that is bounded by and closer to the user-specified precision for any observable. In backend_estimator_v2.py, the _run function is modified to include the sum of the absolute value squared of the coefficients of the observable in the shots estimation.

The following plot shows that the ratio of precision from results to the user-specified precision changes with varying observable term coefficients using the current implementation. In the proposed implementation, this ratio is fixed and close to 1.

![image](https://github.com/Qiskit/qiskit/assets/77505813/f83fcba9-caff-4c02-8735-885c6da14c93)


The plot below shows that in the current implementation, having more than one occurrence of a term in an observable leads to precision results that are different from the user-specified precision due to a fixed number of shots for observables with different numbers of terms and error propagation. In the proposed implementation, the precision from the results remains close to the user-specified precision.



![image](https://github.com/Qiskit/qiskit/assets/77505813/ef82789f-f98a-42e1-bcde-b2067de99937)

jsaroni committed:
Update backend_estimator_v2.py

Improving shot estimate to give errors within user-given precision.

jsaroni committed:
Update backend_estimator_v2.py

Used 1-line replacement for coeff_sq_sum removing pauli. Had to use _preprocess_pub to get the right data type for using .values(). From the EstimatorV2 tutorial example bc_obs is of the form [{'II': 1.0, 'IZ': 2.0, 'XI': 3.0}] so bc_obs[0] gets all the dictionary terms {'II': 1.0, 'IZ': 2.0, 'XI': 3.0}.

jsaroni committed:
Update test_backend_estimator_v2.py

Included a function test_error which shows that for the proposed shot implementation, the user-specified precision is close to the precision from results within the set tolerance by passing the tests. The tests are not passed by the current implementation.

jsaroni committed:
Merge branch 'main' into main

jsaroni committed:
Update backend_estimator_v2.py

Using ravel instead of _preprocess_pub to get coefficients directly as shown by @ihincks

jsaroni committed:
Update test_backend_estimator_v2.py

Changed test function name to test_precision_error and defined multi-term observables to test precision.

jsaroni committed:
Merge branch 'Qiskit:main' into estimator_primitive

SamFerracin commented on qiskit/primitives/backend_estimator_v2.py:
```diff
@@ -161,8 +161,11 @@ def _validate_pubs(self, pubs: list[EstimatorPub]):
     def _run(self, pubs: list[EstimatorPub]) -> PrimitiveResult[PubResult]:
         pub_dict = defaultdict(list)
         # consolidate pubs with the same number of shots
+        coeff_sq_sum = 0
```
I am a little bit confused by the math here..

The variance of a linear combination `aX + bY + ...` of observables `X, Y, ...` is given by:

> `var(aX + bY + ...) = a**2 Var(X) + b**2 Var(Y) + ...`, 


where `Var(O) = 1/N_O` is the variance of the observable `O` and `N_O` is the number of shots for the observable `O`.
In our case, we want `var(aX + bY + ...)` to be equal to `epsilon**2`, and we want to use the same number of shots `N` for every observable `O`.  Therefore, we have: 

> `epsilon**2 = a**2/N + b**2/N + ...`, 


which gives:

> `N = (a**2 + b**2 + ...)/epsilon**2`

So here is how I would code this up:
```
for i, pub in enumerate(pubs):
    coeff_sq_sum = 0
    for obs in pub.observables.ravel():
        coeff_sq_sum += sum(coeff**2 for coeff in obs.values())
    shots = int(math.ceil(coeff_sq_sum / pub.precision**2))
```

Am I missing something?

SamFerracin replied:
FYI, there's a related issue that may shed some light
https://github.com/Qiskit/qiskit-ibm-runtime/issues/1684

jsaroni replied:
Right. This is how it should be as written by @ihincks in https://github.com/Qiskit/qiskit/pull/12618. I have updated it. Given that we are estimating the precision without knowing the target state, setting `Var(O)=1/N_O` bounds the shot number.

ihincks replied:
> Am I missing something?

I think you might be: each element of `pub.observables.ravel()` is itself a multi-term observable. The point of the max was to find the worst offender.

SamFerracin replied:
I see. So every `obs` is not an observable, but a list of (potentially more than one) observables.
@jsaroni could you add a comment about this? It will help us remember in the future

jsaroni replied:
Done.

jsaroni committed:
Update backend_estimator_v2.py

Find coeff_sq_sum for each pub

jsaroni committed:
Update backend_estimator_v2.py

Added comment, "each element of pub.observables.ravel() may be a multi-term observable".

t-imamichi commented on test/python/primitives/test_backend_estimator_v2.py:
```diff
@@ -388,6 +388,29 @@ def test_diff_precision(self, backend, abelian_grouping):
 
+    @combine(backend=BACKENDS, abelian_grouping=[True, False])
+    def test_precision_error(self, backend, abelian_grouping):
+        "Test for error"
```
Could you write a docstring `"""..."""` instead of a string `"..."`?
"Test for error" looks too generic. Could you write a little more details?

jsaroni replied:
How does """Test that errors are within user-specified precision""" sound?

t-imamichi commented:
Could you apply black following https://github.com/Qiskit/qiskit/blob/main/CONTRIBUTING.md#style-and-lint? CI fails due to a style error.

t-imamichi commented on test/python/primitives/test_backend_estimator_v2.py:
```diff
@@ -388,6 +388,29 @@ def test_diff_precision(self, backend, abelian_grouping):
         np.testing.assert_allclose(result[1].data.evs, [1.901141473854881], rtol=self._rtol)
 
+    @combine(backend=BACKENDS, abelian_grouping=[True, False])
+    def test_precision_error(self, backend, abelian_grouping):
```
The test function named "test_precision_error", but there is no error checking in the test.
`test_something_error` usually checks a particular error is actually raised or not by `self.assertRaises`.
The current test looks checking `stds` with the target precision. If it's correct, I would suggest a function name `test_precision_and_stds`, for example.

jsaroni replied:
Ok, changed it to test_precision_and_stds.

jsaroni committed:
Update test_backend_estimator_v2.py

Changing function name to test_precision_and_stds and using docstring """Test that errors are within user-specified precision"""

jsaroni commented:
> Could you apply black following https://github.com/Qiskit/qiskit/blob/main/CONTRIBUTING.md#style-and-lint? CI fails due to a style error.

I used the black formatting. I checked that lint does not return an error locally. I still get the CI error

```
1 file would be reformatted, 1277 files would be left unchanged. 
lint: exit 1 (38.82 seconds) /home/vsts/work/1/s> black --check qiskit test tools examples setup.py pid=13021 
docs: OK (598.62=setup[407.74]+cmd[21.00,169.88] seconds) 
lint: FAIL code 1 (209.40=setup[170.58]+cmd[38.82] seconds) 
evaluation failed :( (808.81 seconds) 
##[error]Bash exited with code '255'. 
```

jsaroni committed:
Merge branch 'main' into estimator_primitive

t-imamichi committed:
black

t-imamichi commented:
No problem. I applied black to this branch. Hopefully CI passes now.

t-imamichi commented:
Could you write a release note following the guide? I think this PR should be "features_primitives" section.
https://github.com/Qiskit/qiskit/blob/main/CONTRIBUTING.md#release-notes

jsaroni committed:
Merge branch 'Qiskit:main' into estimator_primitive

jsaroni committed:
Add files via upload

Adding a release note in the "features_primitives" section.

jsaroni commented:
> Could you write a release note following the guide? I think this PR should be "features_primitives" section. https://github.com/Qiskit/qiskit/blob/main/CONTRIBUTING.md#release-notes

Thanks. I have added a release note for review https://github.com/jsaroni/qiskit/blob/estimator_primitive/releasenotes/notes/backend_estimator_stds-aa6893ab647567f1.yaml

t-imamichi commented on releasenotes/notes/backend_estimator_stds-aa6893ab647567f1.yaml:
```diff
@@ -0,0 +1,29 @@
+      # calculate [ <psi(theta)|hamiltonian|psi(theta)> ]
+      job = estimator.run([(isa_psi, isa_observables, [theta])], precision=0.01)
+      pub_result = job.result()[0]
+      print(f"Expectation values: {pub_result.data.evs}")
```
Thank you for writing a reno. You don't need to write an example because this PR does not change API.
```suggestion
    An estimation of the shots has been updated in :meth:`.BackendEstimatorV2.run` to account for the error from multi-term observables.
```

jsaroni committed:
Update releasenotes/notes/backend_estimator_stds-aa6893ab647567f1.yaml

Removing example since PR does not change API.

Co-authored-by: Takashi Imamichi <31178928+t-imamichi@users.noreply.github.com>

t-imamichi commented on qiskit/primitives/backend_estimator_v2.py:
```diff
@@ -162,7 +162,11 @@ def _run(self, pubs: list[EstimatorPub]) -> PrimitiveResult[PubResult]:
+            for obs in pub.observables.ravel():
+                # each element of pub.observables.ravel() may be a multi-term observable
+                coeff_sq_sum = max(coeff_sq_sum, sum(coeff**2 for coeff in obs.values()))
+            shots = int(math.ceil(coeff_sq_sum / pub.precision**2))
```
If all coefficients are 0, shots is set 0 and it results in an error. You need to set at least 1 shot.
Could you add the following case to the unit test?
```suggestion
            shots = int(math.ceil(coeff_sq_sum / pub.precision**2))
            shots = max(1, shots)
```

The following script causes an error only for this PR.
```python
from qiskit import QuantumCircuit
from qiskit.primitives import BackendEstimatorV2
from qiskit_aer import AerSimulator
from qiskit.quantum_info import SparsePauliOp

backend = AerSimulator()
estimator = BackendEstimatorV2(backend=backend)

qc = QuantumCircuit(1)
qc.h(0)
op = SparsePauliOp.from_list([("X", 0), ("Y", 0)])

result = estimator.run([(qc, op)]).result()
print(result[0].data.evs)
```

this PR
```
qiskit.exceptions.QiskitError: 'No counts for experiment "0"'
```

main branch
```
0.0
```

jsaroni replied:
@t-imamichi I see, thanks. Could I use one line?
`shots = max(1, int(math.ceil(coeff_sq_sum / pub.precision**2)))`

t-imamichi replied:
Sure

jsaroni committed:
Update qiskit/primitives/backend_estimator_v2.py

Setting at least 1 shots to avoid error when all coefficients are 0.

Co-authored-by: Takashi Imamichi <31178928+t-imamichi@users.noreply.github.com>

jsaroni committed:
Update backend_estimator_v2.py

One line shots assignment.

jsaroni committed:
Merge branch 'main' into estimator_primitive

t-imamichi commented:
Could you write a test of zero coefficient based on this script?
https://github.com/Qiskit/qiskit/pull/12711#discussion_r1673645697

jsaroni committed:
Update test_backend_estimator_v2.py

Added a test in test_precision_and_stds for the case when all observable term coefficients are zero.

jsaroni commented:
> Could you write a test of zero coefficient based on this script? [#12711 (comment)](https://github.com/Qiskit/qiskit/pull/12711#discussion_r1673645697)

Ok. I added a test in `test_precision_and_stds` for when all observable term coefficients are zero. In this case both stds and evs are zero regardless of the set precision.

t-imamichi committed:
Merge branch 'main' into estimator_primitive

t-imamichi commented:
Thank you for adding the test. I'm running CI...

jsaroni committed:
Update test_backend_estimator_v2.py

More test cases.

t-imamichi commented:
The recent changes breaks the style. You need to apply black again.

jsaroni committed:
Merge branch 'Qiskit:main' into estimator_primitive

jsaroni committed:
Update test_backend_estimator_v2.py

ihincks commented on test/python/primitives/test_backend_estimator_v2.py:
```diff
@@ -388,6 +388,78 @@ def test_diff_precision(self, backend, abelian_grouping):
         np.testing.assert_allclose(result[1].data.evs, [1.901141473854881], rtol=self._rtol)
 
+    @combine(backend=BACKENDS, abelian_grouping=[True, False])
+    def test_precision_and_stds(self, backend, abelian_grouping):
```
Optional

```suggestion
    @combine(backend=BACKENDS, abelian_grouping=[True, False], hamiltonian=...)
    def test_precision_and_stds(self, backend, abelian_grouping, hamiltonian):
```

ihincks commented on test/python/primitives/test_backend_estimator_v2.py:
```diff
@@ -388,6 +388,78 @@ def test_diff_precision(self, backend, abelian_grouping):
+            )
+            np.testing.assert_array_less(
+                result[1].data.stds, [self._precision] * len(hamiltonian[i])
+            )
```
```suggestion
        for hamiltonian in hamiltonians:
            job = estimator.run(
                [(self.psi[0], hamiltonian, [self.theta[0]])], precision=self._precision
            )
            result = job.result()
            np.testing.assert_array_less(
                result[0].data.stds, [self._precision] * len(hamiltonian)
            )
            # The result of the second run is the same
            job = estimator.run(
                [
                    (self.psi[0], hamiltonian, [self.theta[0]]),
                    (self.psi[0], hamiltonian, [self.theta[0]]),
                ],
                precision=self._precision,
            )
            result = job.result()
            np.testing.assert_array_less(
                result[0].data.stds, [self._precision] * len(hamiltonian)
            )
            np.testing.assert_array_less(
                result[1].data.stds, [self._precision] * len(hamiltonian)
            )
```

ihincks replied:
```suggestion
```

ihincks commented on test/python/primitives/test_backend_estimator_v2.py:
```diff
@@ -388,6 +388,78 @@ def test_diff_precision(self, backend, abelian_grouping):
+        ]
+        for i in range(len(hamiltonian)):
+            job = estimator.run(
+                [(self.psi[0], hamiltonian[i], [self.theta[0]])], precision=self._precision
```
Why is the dimension being increased here? Not a big deal, but don't add it unless there's a reason.
```suggestion
                [(self.psi[0], hamiltonian[i], self.theta[0]]), precision=self._precision
```

jsaroni replied:
Resolved.

ihincks commented on test/python/primitives/test_backend_estimator_v2.py:
```diff
@@ -388,6 +388,78 @@ def test_diff_precision(self, backend, abelian_grouping):
+            )
+            result = job.result()
+            np.testing.assert_array_less(
+                result[0].data.stds, [self._precision] * len(hamiltonian[i])
```
Two issues here:
 1. assert_array_less already does broadcasting, so no need to replicate the precision ourselves
 2. len(hamilotian) for hamiltonian[0] [1] and [2] does not correspond to the size of `.stds`

```suggestion
                result[0].data.stds, self._precision
```

jsaroni replied:
Ok.

ihincks commented on test/python/primitives/test_backend_estimator_v2.py:
```diff
@@ -388,6 +388,78 @@ def test_diff_precision(self, backend, abelian_grouping):
         np.testing.assert_allclose(result[0].data.evs, [1.901141473854881], rtol=self._rtol)
         np.testing.assert_allclose(result[1].data.evs, [1.901141473854881], rtol=self._rtol)
 
+    @combine(backend=BACKENDS, abelian_grouping=[True, False])
```
What is the value of looping over backends and abelian_grouping? Does it actually increase test coverage meaningfully?

jsaroni replied:
I removed the looping abelian_grouping and selected just one backend.

ihincks commented on test/python/primitives/test_backend_estimator_v2.py:
```diff
@@ -388,6 +388,78 @@ def test_diff_precision(self, backend, abelian_grouping):
+            )
+
+    @combine(backend=BACKENDS, abelian_grouping=[True, False])
+    def test0_precision_and_stds(self, backend, abelian_grouping):
```
```suggestion
    def test_precision_with_zero_coefficients(self, backend, abelian_grouping):
```

ihincks commented on test/python/primitives/test_backend_estimator_v2.py:
```diff
@@ -388,6 +388,78 @@ def test_diff_precision(self, backend, abelian_grouping):
+                SparsePauliOp.from_list([("ZZ", 0)]),
+                SparsePauliOp.from_list([("ZZ", 0), ("ZZ", 0)]),
+            ],
+        ]
```
Avoid testing more than necessary
```suggestion
        hamiltonian = [
            SparsePauliOp.from_list([("II", 0), ("IZ", 0), ("XI", 0)]),
            SparsePauliOp.from_list([("ZZ", 0)]),
            [
                SparsePauliOp.from_list([("II", 0), ("IZ", 0), ("XI", 0)]),
                SparsePauliOp.from_list([("ZZ", 1)]),
                SparsePauliOp.from_list([("ZZ", 0), ("ZZ", 0)]),
            ],
        ]
```

ihincks commented on test/python/primitives/test_backend_estimator_v2.py:
```diff
@@ -388,6 +388,78 @@ def test_diff_precision(self, backend, abelian_grouping):
+            [
+                SparsePauliOp.from_list([("II", 0), ("IZ", 0), ("XI", 0)]),
+                SparsePauliOp.from_list([("ZZ", 0)]),
+                SparsePauliOp.from_list([("ZZ", 0), ("ZZ", 0)]),
```
Could we get more test coverage by making one of these non-zero?

jsaroni replied:
Added a zero and non-zero coefficient test case.

jsaroni committed:
Update test/python/primitives/test_backend_estimator_v2.py

Using 'for hamiltonian in hamiltonians' and removing second run.

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

jsaroni committed:
Update test/python/primitives/test_backend_estimator_v2.py

Changing function name to exclude int.

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

jsaroni committed:
Update test/python/primitives/test_backend_estimator_v2.py

Removing unnecessary tests.

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

jsaroni committed:
Update test_backend_estimator_v2.py

Adding zero and non-zero coefficient test case.

jsaroni committed:
Update test_backend_estimator_v2.py

Increasing lint rating. Applied black.

jsaroni committed:
Merge branch 'estimator_primitive' of https://github.com/jsaroni/qiskit into estimator_primitive

jsaroni committed:
Update test_backend_estimator_v2.py

Using list with dictionaries for observables.

jsaroni committed:
Update test_backend_estimator_v2.py

remove @combine backend

jsaroni committed:
Update test_backend_estimator_v2.py

transpilation needed for backend using RealAmplitudes.

jsaroni committed:
Update test_backend_estimator_v2.py

t-imamichi commented:
Could you apply black and check lint too?

jsaroni committed:
Update test_backend_estimator_v2.py

jsaroni committed:
Update test_backend_estimator_v2.py

jsaroni committed:
Update test_backend_estimator_v2.py

jsaroni committed:
Update test_backend_estimator_v2.py

jsaroni committed:
Update test_backend_estimator_v2.py

jsaroni committed:
Update test_backend_estimator_v2.py

Applied black.

jsaroni committed:
Update test_backend_estimator_v2.py

Linting

jsaroni commented:
> Could you apply black and check lint too?

Applied it.

t-imamichi commented:
Could you replied to Ian's questions and suggestions?

jsaroni commented:
> Could you replied to Ian's questions and suggestions?

Replied. Resolved the posted comments.

t-imamichi commented:
Could you resolve the conflict?

jsaroni commented:
> Could you resolve the conflict?

It shows this section in red. I'm not sure why.
![sshot](https://github.com/user-attachments/assets/cb80eb59-5e3d-40fe-ae51-3ff30dc0ca90)

t-imamichi commented:
That's the conflict part you need to resolve. Could you take a look at the documentation? https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/addressing-merge-conflicts/resolving-a-merge-conflict-on-github

jsaroni committed:
Merge branch 'main' into estimator_primitive

t-imamichi reviewed (approved):
LGTM

t-imamichi committed:
Merge branch 'main' into estimator_primitive

jsaroni committed:
Merge branch 'main' into estimator_primitive

jsaroni committed:
Merge branch 'Qiskit:main' into estimator_primitive
